pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -908,9 +924,49 @@ contract StablePhantomPool is
 
     // Store the latest invariant based on the adjusted balances after the join or exit, using current rates.
     // Also cache the amp factor, so that the invariant is not affected by amp updates between joins and exits.
-    function _updateInvariantAfterJoinExit(uint256 currentAmp, uint256[] memory balancesWithoutBpt) internal {
+    function _updateInvariantAfterJoinExit(
+        uint256 currentAmp,
+        uint256[] memory balancesWithoutBpt,
+        uint256 preJoinExitInvariant,
+        uint256 virtualSupply,
+        uint256 bptAmount
+    ) internal {
+        uint256 postJoinExitInvariant = StableMath._calculateInvariant(currentAmp, balancesWithoutBpt);
+
+        // Compute the growth ratio between the pre- and post-join/exit balances.
+        // Note that the pre-join/exit invariant is *not* the invariant from the last join,
+        // but computed from the balances before this particular join/exit.
+
+        uint256 protocolSwapFeePercentage = getProtocolFeePercentageCache(ProtocolFeeType.SWAP);
+
+        if (protocolSwapFeePercentage > 0) {
+            uint256 invariantGrowthRatio = postJoinExitInvariant > preJoinExitInvariant
+                ? postJoinExitInvariant.divDown(preJoinExitInvariant)
+                : preJoinExitInvariant.divDown(postJoinExitInvariant);
+
+            // Compute the bptIncrease = (1 + bptAmount/virtualSupply)
+            uint256 bptIncreaseRatio = FixedPoint.ONE.add(bptAmount.divDown(virtualSupply));
+
+            // The difference between the invariant growth and bpt increase rates must be due to the
+            // balance change from this join/exit.
+            // Protocol fees due = (invariant growth / bpt increase - 1) * virtual supply * protocol fee %
+            // For instance, if the invariant growth is 1.05, and the bpt increase is 1.0475, with 1000 supply,
+            // and a protocol fee of 50%, we would mint (1.05/1.0475 - 1) * 1000 * 0.5 = 1.193 BPT.
+
+            if (invariantGrowthRatio > bptIncreaseRatio) {
+                uint256 protocolFeeAmount = invariantGrowthRatio
+                    .divDown(bptIncreaseRatio)
+                    .sub(FixedPoint.ONE)
+                    .mulDown(virtualSupply)
+                    .mulDown(protocolSwapFeePercentage);
```

Comment:
I'm having some trouble with this, I'll try to derive it via an alternative method and get back to you.

Reply:
This can be addressed later once we have tests.